opm/simulators/flow/ConvergenceOutputConfiguration.cpp

```diff
@@ -0,0 +1,115 @@
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <opm/simulators/flow/ConvergenceOutputConfiguration.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <regex>
+#include <stdexcept>
+#include <string_view>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace {
+    std::vector<std::string> tokenizeOptionValues(std::string_view options)
+    {
+        const auto split = std::regex { R"(\s*,\s*)" };
+        return {
+            std::cregex_token_iterator {
+                options.begin(), options.end(), split, -1
+            },
+            std::cregex_token_iterator {}
+        };
+    }
+
+    void reportUnsupportedOptionValuesAndThrow(std::vector<std::string> unsupp)
+    {
+        std::sort(unsupp.begin(), unsupp.end());
+        auto u = std::unique(unsupp.begin(), unsupp.end());
+
+        const auto pl = (std::distance(unsupp.begin(), u) != 1) ? "s" : "";
+
+        throw std::invalid_argument {
+            fmt::format("Unsupported convergence output "
+                        "option value{}: {}\n"
+                        "Supported values are \"none\", "
+                        "\"steps\", and \"iterations\"", pl,
+                        fmt::join(unsupp.begin(), u, ", "))
+        };
+    }
+
+    std::vector<Opm::ConvergenceOutputConfiguration::Option>
+    getOptions(std::string_view options)
+    {
+        using Option = Opm::ConvergenceOutputConfiguration::Option;
+
+        auto opt = std::vector<Option>{};
+
+        const auto values = std::unordered_map<std::string, Option> {
+            { "none"      , Option::None       },
+            { "step"      , Option::Steps      }, // Alias for 'steps' (plural)
+            { "steps"     , Option::Steps      },
+            { "iteration" , Option::Iterations }, // Alias for 'iterations' (plural)
+            { "iterations", Option::Iterations },
+        };
+
+        auto unsupp = std::vector<std::string>{};
+        for (const auto& value : tokenizeOptionValues(options)) {
+            if (auto option = values.find(value); option != values.end()) {
+                opt.push_back(option->second);
+            }
+            else {
+                unsupp.push_back(value);
+            }
+        }
+
+        if (! unsupp.empty()) {
+            reportUnsupportedOptionValuesAndThrow(std::move(unsupp));
```

Comment:
Do you use some kind of guideline on when to use move and when not? What exactly will it buy you here?

Seems like this code is not performance critical and one code rely on inlining. It seems less readable at least,

Reply:
> Do you use some kind of guideline on when to use move and when not?

Yes.  I write `std::move()` when at least one of the following two conditions is satisfied

1. When I have to, e.g., because the function being called takes a value of a non-copyable type (e.g., a `std::unique_ptr<>`) or when the function being called takes an r-value and I only have an object
2. When `std::move()` expresses the desired semantics, usually "I'm transferring ownership because I don't need it anymore."  While `std:move()` is **primarily** an optimisation of copy, that's not **all** that it does.

> What exactly will it buy you here?

In this case it expresses the "I'm transferring ownership of this object to the other function because I don't need it anymore."  The function being called needs a separate vector object because it's going to manipulate it in order to do its job and I have one of those objects at the call site.  For what it's worth, we do the same in [`Summary.cpp`](https://github.com/OPM/opm-common/blob/9d4b36a891834e537881e788d731251002fc2776/src/opm/output/eclipse/Summary.cpp#L4095-L4096), and for the exact same reason.